dbms/src/DataStreams/WindowTransformAction.cpp

```diff
@@ -1506,26 +1561,37 @@ void WindowTransformAction::tryCalculate()
             assert(frame_started);
             assert(frame_ended);
 
-            if (window_description.need_decrease)
+            if constexpr (need_decrease)
                 updateAggregationState<true>();
             else
                 updateAggregationState<false>();
 
-            // Write out the results.
-            // TODO execute the window function by block instead of row.
-            writeOutCurrentRow();
+
+            if constexpr (has_shortcut)
+            {
+                // When we reach here, partition must be ended
+                assert(partition_ended);
+                writeBatchResult();
```

Comment:
how about return `tmp_row` to update `current_row`? which is more clear.

Reply:
> how about return `tmp_row` to update `current_row`? which is more clear.

done